src/BenchmarkDotNet/Toolchains/CsProj/CsProjGenerator.cs

```diff
@@ -246,36 +247,32 @@ private static string GetIndentedXmlString(XmlDocument doc)
         /// returns a path to the project file which defines the benchmarks
         /// </summary>
         [PublicAPI]
-        protected virtual FileInfo GetProjectFilePath(Type benchmarkTarget, ILogger logger)
+        protected virtual FileInfo GetProjectFilePath(BenchmarkCase benchmark, ILogger logger)
         {
+            var benchmarkTarget = benchmark.Descriptor.Type;
+
             if (!GetSolutionRootDirectory(out var rootDirectory) && !GetProjectRootDirectory(out rootDirectory))
             {
                 logger.WriteLineError(
                     $"Unable to find .sln or .csproj file. Will use current directory {Directory.GetCurrentDirectory()} to search for project file. If you don't use .sln file on purpose it should not be a problem.");
                 rootDirectory = new DirectoryInfo(Directory.GetCurrentDirectory());
             }
 
-            // important assumption! project's file name === output dll name
-            string projectName = benchmarkTarget.GetTypeInfo().Assembly.GetName().Name;
+            List<string> notFound = new List<string>();
+            foreach (ILocator locator in benchmark.Config.GetLocators())
+            {
+                if (locator.LocatorType != LocatorType.ProjectFile)
+                    continue;
 
-            var possibleNames = new HashSet<string> { $"{projectName}.csproj", $"{projectName}.fsproj", $"{projectName}.vbproj" };
-            var projectFiles = rootDirectory
-                .EnumerateFiles("*proj", SearchOption.AllDirectories)
-                .Where(file => possibleNames.Contains(file.Name))
-                .ToArray();
+                var path = locator.Locate(rootDirectory, benchmarkTarget);
 
-            if (projectFiles.Length == 0)
-            {
-                throw new NotSupportedException(
-                    $"Unable to find {projectName} in {rootDirectory.FullName} and its subfolders. Most probably the name of output exe is different than the name of the .(c/f)sproj");
-            }
-            else if (projectFiles.Length > 1)
-            {
-                throw new NotSupportedException(
-                    $"Found more than one matching project file for {projectName} in {rootDirectory.FullName} and its subfolders: {string.Join(",", projectFiles.Select(pf => $"'{pf.FullName}'"))}. Benchmark project names needs to be unique.");
+                if (path.Exists)
+                    return path;
+
+                notFound.Add(path.FullName);
             }
 
-            return projectFiles[0];
+            throw new NotSupportedException($"Unable to find project file in {rootDirectory}. Attempted location(s): " + string.Join(", ", notFound));
```

Comment:
I know `NotSupportedException` was used previously, but it looks weird to me. I think it should be `FileNotFoundException` instead.

Reply:
Agree